### PR TITLE
Use string.Template in RosoutStreamHandler

### PR DESCRIPTION
**Pull request by mgrrx**
This allows to use e.g. $message instead of ${message} in ROSCONSOLE_FORMAT.

### CODE REVIEW

File: tools/rosgraph/src/rosgraph/roslogging.py [hunk 1]
```diff
@@ -232,31 +233,32 @@ def __init__(self, colorize=True, stdout=None, stderr=None):
...
+        msg = self._template.substitute(severity=level,
+                                        message=record_message,
+                                        walltime='%f' % time.time(),
+                                        thread=record.thread,
+                                        logger=record.name,
+                                        file=record.pathname,
+                                        line=record.lineno,
+                                        function=record.funcName,
+                                        node=node_name,
+                                        time=time_str)
```
Comment (mikepurvis) [inline tools/rosgraph/src/rosgraph/roslogging.py:262]: Slight preference here for creating a `replacements` dict and then doing

```
msg = self._template.substitute(**replacements)
```

But I don't think it much matters, and what you have is more succinct.

**Review by mikepurvis — APPROVED** [review]
Seems like a good change to me; would be great to have tests that cover these log expansions somehow, but whatever.

**Comment by dirk-thomas** [comment]
I am not sure if that is likely but if an existing format string contains a dollar sign which isn't followed by a variable name in curly braces this patch will make this use case break. While the chances are probably low I am not sure this patch adds a lot of valuable on the other side for this chance of breakage.

**Comment by mgrrx** [comment]
My intention here was to simplify the existing code:
- iteratively replacing substrings by values seems like a bad idea to me. And since string.Template uses regular expressions it is also faster
- I moved the os.environ.get part into the constructor since the environment variables should not change during run time.

**Comment by dirk-thomas** [comment]
> I am not sure if that is likely but if an existing format string contains a dollar sign which isn't followed by a variable name in curly braces this patch will make this use case break. While the chances are probably low I am not sure this patch adds a lot of valuable on the other side for this chance of breakage.

@mgrrx Can you please respond to my earlier concern.

Until then I will mark this for future distros to avoid potential regressions with existing uses.

File: tools/rosgraph/src/rosgraph/roslogging.py [hunk 2]
```diff
@@ -233,31 +234,32 @@ def __init__(self, colorize=True, stdout=None, stderr=None):
         except ImportError:
             self._get_time = None
             self._is_wallclock = None
+        self._template = string.Template(os.environ.get(
+            'ROSCONSOLE_FORMAT', '[${severity}] [${time}]: ${message}') + '\n')
```
Comment (dirk-thomas) [inline tools/rosgraph/src/rosgraph/roslogging.py:238]: In the case where the format string is read from the environment variable I would expect it to not have a trailing newline?

File: tools/rosgraph/src/rosgraph/roslogging.py [hunk 3]
```diff
@@ -233,31 +234,32 @@ def __init__(self, colorize=True, stdout=None, stderr=None):
...
             node_name = get_name()
         except ImportError:
             node_name = '<unknown_node_name>'
-        msg = msg.replace('${node}', node_name)
+
         time_str = '%f' % time.time()
         if self._get_time is not None and not self._is_wallclock():
             time_str += ', %f' % self._get_time()
-        msg = msg.replace('${time}', time_str)
-        msg += '\n'
```
Comment (mgrrx) [inline tools/rosgraph/src/rosgraph/roslogging.py:260]: @dirk-thomas answering your question: because of this \n here ;)

Comment (dirk-thomas) [reply tools/rosgraph/src/rosgraph/roslogging.py:260]: But this line is being removed.

Comment (mgrrx) [reply tools/rosgraph/src/rosgraph/roslogging.py:260]: yep this line is removed and instead I added the \n in line 238

Comment (dirk-thomas) [reply tools/rosgraph/src/rosgraph/roslogging.py:260]: Nvm, the trailing `\n` is *outside* of the get env call.

**Comment by dirk-thomas** [comment]
@mgrrx Please rebase to target `noetic-devel`.

**Comment by dirk-thomas** [comment]
@mgrrx Friendly ping.

**Comment by dirk-thomas** [comment]
> Until then I will mark this for future distros to avoid potential regressions with existing uses.

With Noetic being released and likely being the last ROS 1 distro and this ticket proposing to change behavior which shouldn't target an already released distro I will go ahead and close it.

**Pull request closed without merging**